### DISCUSSION TITLE
Obstacle avoidance with Flow and Multiranger deck

### DISCUSSION THREAD

**Discussion by ADW-2024** [bitcraze/discussions#1268] [Q&A]
Hello everyone,
Ive been trying to get the crazyflie to take off and move to a location 2 meters forward if it does not detect an obstacle and move to either side if it does detect it. I was successful with just using the multiranger and motion commander libraries but I need it to move back and be directly 2 meters forward. I tried to use the HL commander library but it would take off and land without executing the movement. Thinking that it wasn't really needing the HL commander to do what I needed I commented it out and ran it but its doing the same thing. I have been scratching my head for a couple of days now and am thoroughly stuck. In the original run I used a counter in the the loop but tried to change it to a position condition. Here is the code I have so far.

```python
import logging
import sys
import time
import cflib.crtp
from cflib.crazyflie import Crazyflie
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie
from cflib.positioning.motion_commander import MotionCommander
from cflib.utils.multiranger import Multiranger
#from cflib.positioning.position_hl_commander import PositionHlCommander

URI = 'radio://0/100/2M/E7E7E7E703'

logging.basicConfig(level=logging.ERROR)

DEFAULT_HEIGHT = 0.6

VELOCITY = 0.2
velocity_x = None
velocity_y = None
x = 0.0
y = 0.0
z = 0.0

def is_close(range):
    MIN_DISTANCE = 0.4  # m
    if range is None:
        return False
    else:
        return range < MIN_DISTANCE

        
if __name__ == '__main__':
    # Initialize the low-level drivers
    cflib.crtp.init_drivers(enable_debug_driver=False)
    cf = Crazyflie(rw_cache='./cache')
    with SyncCrazyflie(URI, cf=cf) as scf:
      with Multiranger(scf) as multiranger:
        with MotionCommander(scf, default_height=DEFAULT_HEIGHT) as mc: 
            #with PositionHlCommander(scf, default_velocity = VELOCITY, default_height=DEFAULT_HEIGHT) as HLC:
                #HLC.get_position
            while x != 2 and y != 0:
                if not is_close(multiranger.front):   
                    velocity_x = VELOCITY
                    velocity_y = 0.0
                    x = x + VELOCITY
                if is_close(multiranger.front) and not is_close(multiranger.right):
                    velocity_x = 0.0
                    velocity_y = -0.2
                    y = y + velocity_y
                if is_close(multiranger.front) and is_close(multiranger.right):
                    velocity_x = 0.0
                    velocity_y = 0.2
                    y = y + velocity_y
                if y > 0:
                    velocity_y = -0.2
                    y = y + velocity_y
                if y < 0:
                    velocity_y = 0.2
                    y = y + velocity_y
                print("position x =" + str(x) + "and position y =" + str(y))
                mc.start_linear_motion (velocity_x, velocity_y, 0)
                time.sleep(1)
            velocity_x = 0.0    
```

Can someone please help me figure out why my while loop isn't executing?

**Comment by gemenerik** [comment]
I think it is because you initialize `y` as 0, which immediately breaks the while loop (`while ... y != 0:`).

For safety, I would recommend breaking the while loop using conditions such as greater than or less than (`>`, `<`), to prevent scenarios where changing the `VELOCITY` might prevent `x` from reaching exactly 2 and `y` from reaching exactly 0.

Please put your code in a [code block](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks) next time; indentation is important to interpret Python code.

**Reply by ADW-2024** [reply]
Thank you for your reply. I didnt know how to get the code block to generate so your link was very helpful and I will make sure to do that in the future. Im going to be testing the solution late today so fingers crossed I can get it to work.
